webknossos/webknossos/skeleton/entities.py

```diff
@@ -0,0 +1,440 @@
+        self.nx_graph = nx.Graph()
+        if self._enforced_id is not None:
+            self._id = self._enforced_id
+        else:
+            self._id = self._nml.element_id_generator.__next__()
+
+    @property
+    def id(self) -> int:
+        return self._id
+
+    def get_nodes(self) -> List[Node]:
+
+        return [node_view[1] for node_view in self.nx_graph.nodes(data="obj")]
+
+    def get_node_positions(self) -> np.ndarray:
+        return np.array([node.position for node in self.get_nodes()])
+
+    def get_node_by_id(self, node_id: int) -> Node:
+
+        return self.nx_graph.nodes[node_id]["obj"]
+
+    def has_node_id(self, node_id: int) -> bool:
+
+        return node_id in self.nx_graph.nodes
+
+    def add_node(
+        self,
+        position: Vector3,
+        comment: Optional[str] = None,
+        radius: Optional[float] = None,
+        rotation: Optional[Vector3] = None,
+        inVp: Optional[int] = None,
+        inMag: Optional[int] = None,
+        bitDepth: Optional[int] = None,
+        interpolation: Optional[bool] = None,
+        time: Optional[int] = None,
+        is_branchpoint: bool = False,
+        branchpoint_time: Optional[int] = None,
+        _enforced_id: Optional[int] = None,
+        _nml: Optional["Skeleton"] = None,
+    ) -> Node:
+        node = Node(
+            position=position,
+            comment=comment,
+            radius=radius,
+            rotation=rotation,
+            inVp=inVp,
+            inMag=inMag,
+            bitDepth=bitDepth,
+            interpolation=interpolation,
+            time=time,
+            is_branchpoint=is_branchpoint,
+            branchpoint_time=branchpoint_time,
+            enforced_id=_enforced_id,
+            nml=_nml or self._nml,
+        )
+        self.nx_graph.add_node(node.id, obj=node)
+        return node
+
+    def add_edge(self, node_1: Union[int, Node], node_2: Union[int, Node]) -> None:
+
+        id_1 = node_1.id if isinstance(node_1, Node) else node_1
+        id_2 = node_2.id if isinstance(node_2, Node) else node_2
+        self.nx_graph.add_edge(id_1, id_2)
+
+    def get_max_node_id(self) -> int:
+        return max((node.id for node in self.get_nodes()), default=0)
+
+
+@attr.define()
+class Skeleton:
+    """
+    Contains groups and skeletons.
+    """
+
+    name: str
+    scale: Vector3
+    offset: Optional[Vector3] = None
+    time: Optional[int] = None
+    edit_position: Optional[Vector3] = None
+    edit_rotation: Optional[Vector3] = None
+    zoom_level: Optional[float] = None
+    task_bounding_box: Optional[IntVector6] = None
+    user_bounding_boxes: Optional[List[IntVector6]] = None
+
+    root_group: Group = attr.ib(init=False)
+    element_id_generator: Iterator[int] = attr.ib(init=False)
+
+    def __attrs_post_init__(self) -> None:
+        self.element_id_generator = itertools.count()
+        self.root_group = Group(name="Root", children=[], nml=self, is_root_group=False)
```

Comment:
Not sure if correct/incorrect, but I stumbled upon it just now: `self.root_group` ist initialized as `is_root_group=False`... this feels wrong ;-)

Reply:
This was incorrect before, but is completely removed, since the Skeleton now inherits from Group directly.